flink-runtime/src/main/java/org/apache/flink/runtime/rest/handler/legacy/backpressure/BackPressureRequestCoordinator.java

```diff
@@ -0,0 +1,365 @@
+			PendingBackPressureRequest pending = pendingRequests.get(requestId);
+
+			if (pending != null) {
+				pending.collectBackPressureStats(executionId, taskBackPressureRatio);
+
+				// Publish the request result
+				if (pending.isComplete()) {
+					pendingRequests.remove(requestId);
+					rememberRecentRequestId(requestId);
+
+					pending.completePromiseAndDiscard();
+				}
+			} else if (recentPendingRequests.contains(requestId)) {
+				if (LOG.isDebugEnabled()) {
+					LOG.debug("Received late back pressure request {} result of task {}.",
+							requestId, executionId);
+				}
+			} else {
+				if (LOG.isDebugEnabled()) {
+					LOG.debug("Unknown request ID " + requestId);
+				}
+			}
+		}
+	}
+
+	private void rememberRecentRequestId(int requestId) {
+		if (recentPendingRequests.size() >= NUM_GHOST_REQUEST_IDS) {
+			recentPendingRequests.removeFirst();
+		}
+		recentPendingRequests.addLast(requestId);
+	}
+
+	@VisibleForTesting
+	int getNumberOfPendingRequests() {
+		synchronized (lock) {
+			return pendingRequests.size();
+		}
+	}
+
+	// ------------------------------------------------------------------------
+
+	/**
+	 * A pending task back pressure request, which collects task back pressure
+	 * ratio and owns a {@link BackPressureStats} promise.
+	 *
+	 * <p>Access pending request in lock scope.
+	 */
+	private static class PendingBackPressureRequest {
+
+		private final int requestId;
+		private final long startTime;
+		private final Set<ExecutionAttemptID> pendingTasks;
+		private final Map<ExecutionAttemptID, Double> backPressureRatios;
+		private final CompletableFuture<BackPressureStats> backPressureStatsFuture;
+
+		private boolean isDiscarded;
+
+		PendingBackPressureRequest(
+				int requestId,
+				ExecutionAttemptID[] tasksToCollect) {
+
+			this.requestId = requestId;
+			this.startTime = System.currentTimeMillis();
+			this.pendingTasks = new HashSet<>(Arrays.asList(tasksToCollect));
+			this.backPressureRatios = Maps.newHashMapWithExpectedSize(tasksToCollect.length);
+			this.backPressureStatsFuture = new CompletableFuture<>();
+		}
+
+		boolean isComplete() {
+			if (isDiscarded) {
+				throw new IllegalStateException("Discarded");
+			}
+
+			return pendingTasks.isEmpty();
+		}
+
+		void discard(Throwable cause) {
+			if (!isDiscarded) {
+				pendingTasks.clear();
+				backPressureRatios.clear();
+
+				backPressureStatsFuture.completeExceptionally(new RuntimeException("Discarded", cause));
+
+				isDiscarded = true;
+			}
+		}
+
+		void collectBackPressureStats(ExecutionAttemptID executionId, double backPressureRatio) {
+			if (isDiscarded) {
+				throw new IllegalStateException("Discarded");
+			}
+
+			if (pendingTasks.remove(executionId)) {
+				backPressureRatios.put(executionId, backPressureRatio);
+			} else if (isComplete()) {
+				throw new IllegalStateException("Completed");
+			} else {
+				throw new IllegalArgumentException("Unknown task " + executionId);
+			}
+		}
+
+		void completePromiseAndDiscard() {
```

Comment:
```
                          checkCompleted();

			long endTime = System.currentTimeMillis();
			BackPressureStats backPressureStats = new BackPressureStats(
				requestId,
				startTime,
				endTime,
				backPressureRatios);

			backPressureStatsFuture.complete(backPressureStats);
			isDiscarded = true;
```

Reply:
Maybe checkCompleted is not needed, because before this method is called, ```if (pending.isComplete())``` is checked.

Reply:
But you actually checked it in this method via `else` clause. If more than two places use the check complete/discard state, it is better to extract them separately and then use it beforehand to avoid `if...else` clause.